buf/internal/repo.bzl

```diff
@@ -0,0 +1,81 @@
+"""Defines buf_dependencies repo rule"""
+
+load("@bazel_tools//tools/build_defs/repo:utils.bzl", "update_attrs")
+
+_DOC = """
+The `buf_dependencies` rule downloads and generates `proto_library` targets for buf modules hosted on the BSR.
+"""
+
+def _executable_extension(ctx):
+    extension = ""
+    if ctx.os.name.startswith("windows"):
+        extension = ".exe"
+    return extension
+
+def _valid_pin(pin):
+    module_commit = pin.split(":")
+    if len(module_commit) != 2:
+        return False
+
+    if not module_commit[1].isalnum():
+        return False
+
+    module_parts = module_commit[0].split("/")
+    if len(module_parts) != 3:
+        return False
+
+    return True
+
+def _buf_dependencies_impl(ctx):
+    buf = ctx.path(Label("@{}//:buf{}".format(ctx.attr.toolchain_repo, _executable_extension(ctx))))
+
+    for pin in ctx.attr.deps:
+        if not _valid_pin(pin):
+            fail("failed to parse dep should be in the form of <remote>/<owner>/<repo>:<commit>")
+
+        res = ctx.execute(
+            [buf, "export", pin, "--exclude-imports", "--output", ctx.path("")],
+            quiet = False,
+        )
+        if res.return_code != 0:
+            fail("failed with code: {}, error: {}".format(res.return_code, res.stderr))
+
+    ctx.file("WORKSPACE", "workspace(name = \"{name}\")".format(name = ctx.name), executable = False)
+
+    # Run gazelle to generate `proto_library` rules.
+    # Note that this doesn't require the `buf` extension
+    # as all the files are exported to workspace root.
+    #
+    # Copied from `go_repository` rule
+    _gazelle = "@bazel_gazelle_go_repository_tools//:bin/gazelle{}".format(_executable_extension(ctx))
+    gazelle = ctx.path(Label(_gazelle))
+    cmd = [
+        gazelle,
+        "-lang",
+        "proto",
+        "-mode",
+        "fix",
+        "-repo_root",
+        ctx.path(""),
+    ]
+    res = ctx.execute(cmd, quiet = False)
+    if res.return_code != 0:
+        fail("failed with code: {}, error: {}".format(res.return_code, res.stderr))
+
+    return update_attrs(ctx.attr, ["deps"], {})
+
+buf_dependencies = repository_rule(
```

Comment:
I am not fan of the name `buf_dependencies`. Any suggestions?

Reply:
For consistency with the `deps` key in the configuration, we could shorten it to `buf_deps`. We'll eventually have the concept of plugin dependencies (via `buf.plugin.yaml`), so we might want this to be `buf_module_deps`, but that's a little busy too.

We could always reserve `buf_deps` for this rule, then roll out a `buf_plugin_deps` if we ever needed it. I'm not even sure that we'll ever need a rule like `buf_plugin_deps` that for `buf.plugin.yaml` definitions in Bazel, so it's probably a little overkill anyway.

TL;DR I'd probably just name this `buf_deps`.

Reply:
Ah I spoke too earlier - I see we already have a `buf_deps` rule used elsewhere. If it's a matter of which one is more user facing, we should make the user-facing one `buf_deps`, but it looks like that's already the case.

Reply:
How about `buf_repository` like `git_repository`, and `local_repository` and renaming the `deps` attribute to `modules`?

Reply:
`buf_repository` seems on face value to be inconsistent with our naming, same with `deps` to `modules`. Having `buf_deps` and `buf_dependencies` worries me too. I'd finish the docs first, and then implement what makes most sense from a consistency perspective.

Reply:
So `deps` can't be used as an attr, cause things like aspects and bazel in general expects them to be labels, see https://github.com/bufbuild/rules_buf/pull/14#discussion_r898269203. I changed the attr to `modules`

Reply:
The canonical example for using an auto updated macro to track deps is in https://github.com/bazelbuild/bazel-gazelle#update-repos, which works something like
```
# repositories.bzl
def go_repositories():
    # The go_repository rules below get updated by running:
    #   $ gazelle update-repos -from_file=go.work -to_macro=repositories.bzl%go_repositories
    go_repository(...)
    go_repository(...)

# WORKSPACE
load("//:repositories.bzl", "go_repositories")

go_repositories() # this loads all the go_repository rules here into the workspace
```

When loading dependencies provided by external rulesets, the convention is to have something like a `_dependencies` in a `deps.bzl`:
```
load("@io_bazel_rules_go//go:deps.bzl", "go_rules_dependencies")
go_rules_dependencies()
load("@bazel_gazelle//:deps.bzl", "gazelle_dependencies")
gazelle_dependencies()
```
However, our usage of `buf_deps` in this PR is not a rule provided by rules_buf, but a macro name used for the current WORKSPACE's dependencies in one of the examples. As far as I know, there are no conventions for that other than maybe `repositories.bzl%go_repositories` mentioned in the gazelle documentation (usually others should not be importing your rules' development dependencies).

However, in our case the current `buf_dependencies` rule maps to downloading a list of repositories for a `buf.work`/`buf.lock`. This is unlike the more granular repo rule you'd see in bazel such as the `go_repository` rule which would only download an individual go module (and you'd generate many of those per go.mod, unlike the current 1 buf_dependencies per buf.lock/buf.work).

If I recall, the problem with doing a rule per buf module approach is with workspaces. Different workspace modules can have different versions of a common dependency pinned, so either we'd have to put versions into rule names or we couldn't replicate the cli build behavior.

--- 

Given the above, a `buf_dependencies` rule that has multiple `module` seems the most straightforward to replicate workspaces in bazel.

One last thing to note is that since bazel does put a premium on build reproducibility, dependency rules should generally have some digest in them for making sure the content received for a pointer is the same when the rule was committed as in the future. `go_repository` has `sum` for this, `http_archive` uses `sha256`. Ideally we'd have something like this for `buf_dependencies`, but with the sums being absent from buf.lock I don't think we can really do much here without first addressing that in the cli itself.